### PR TITLE
Implement new web socket library alternative to socket.io

### PR DESCRIPTION
**Pull request by tsbraun1891**

### CODE REVIEW

**Comment by p-flock** [comment]
disconnection bug fixed - still need to figure out why bgw is breaking on this branch only

File: lib/jiff-client.js [hunk 1]
```diff
@@ -46,6 +46,7 @@ var counters = require('./client/arch/counters.js');
 var guardedSocket = require('./client/socket/mailbox.js');
 var internalSocket = require('./client/socket/internal.js');
 var socketEvents = require('./client/socket/events.js');
+var jiff_websockets = require('./ext/jiff-client-websockets');
```
Comment (p-flock) [inline lib/jiff-client.js:49]: remove

File: lib/jiff-client.js [hunk 2]
```diff
@@ -111,6 +112,7 @@ function JIFFClient(hostname, computation_id, options) {
   if (!this.hostname.endsWith('/')) {
     this.hostname = this.hostname + '/';
   }
+  this.jiff_websockets = jiff_websockets;
```
Comment (p-flock) [inline lib/jiff-client.js:115]: remove

File: tests/suite/config/websocket/constant arithmetic.json [hunk 3]
```diff
@@ -0,0 +1,79 @@
+{
+  "tests": [ "+", "-", "*", "|", "^", "cdivfac", "/Zp127", "/Zp2039", "cpow-acc", "cpow-inacc" ],
+  "suiteConf": {
+    "port": 3001,
+    "extensions": [ "websocket" ],
+    "generation": {
+      "file": "config/websocket/generation.js",
```
Comment (p-flock) [inline tests/suite/config/websocket/constant arithmetic.json:7]: point to default generation.

File: tests/suite/config/websocket/ifelse.json [hunk 4]
```diff
@@ -0,0 +1,26 @@
+{
+  "tests": [ "sif_else", "s1if_else", "s2if_else", "cif_else" ],
+  "suiteConf": {
+    "port": 3001,
+    "extensions": [ "websocket" ],
+    "generation": {
+      "file": "config/websocket/generation.js",
```
Comment (p-flock) [inline tests/suite/config/websocket/ifelse.json:7]: point to default generation

File: tests/suite/server.js [hunk 5]
```diff
@@ -7,12 +7,15 @@ var http = require('http').Server(app);
...
 var jiffRestAPIServer =  require('../../lib/ext/jiff-server-restful');
 var jiffBigNumberServer = require('../../lib/ext/jiff-server-bignumber');
+var jiffWebSocketServer = require('../../lib/ext/jiff-server-websockets');
 
 var options = {
   logs: true,
   app: app
 };
 
+console.log("Extensions: " + extensions);
```
Comment (p-flock) [inline tests/suite/server.js:17]: remove log

**Review by p-flock — CHANGES_REQUESTED** [review]
All looks good after comments are addressed. Add distribution details to extensions documentation md file.

Final thing after these changes is confirming the __internal socket functionality.

**Comment by p-flock** [comment]
changing the module format for the websocket extensions works in browserify / web-browsers, but breaks in node - a future PR should change all the extensions' module formats and change how we apply extensions to function in node as well.

**Review by p-flock — APPROVED** [review]
All looks good, the  __internal socket seems to work with the extension. I will update the dist/ files and then it is good to go.

File: lib/ext/jiff-client-restful.js [hunk 6]
```diff
@@ -192,6 +192,9 @@
     jiff.maxBatchSize = (options.maxBatchSize == null) ? 150 : options.maxBatchSize;
 
     // Stop the socket just in case it got connected somehow (if user forgot to disabled autoConnect)
+    if (jiff.socket == null) {
+      jiff.socketConnect(jiff);
+    }
```
Comment (KinanBab) [inline lib/ext/jiff-client-restful.js:197]: not sure I understand why we have this line here

File: lib/ext/jiff-client-websockets.js [hunk 7]
```diff
@@ -0,0 +1,351 @@
+/**
```
Comment (KinanBab) [inline lib/ext/jiff-client-websockets.js:1]: this looks good to me

**Comment by KinanBab** [comment]
Overall this is very nice. I am happy that we have websockets now, plus that it was implemented with a relatively "brief"/clean code. I was worried it would end up re-implementing and duplicating many things! Nice work folks.

**Comment by p-flock** [comment]
> This looks good to me, two minor comments:
> 
>     1. Unsure if the extension will work properly with a server that participates in computation. The reason I worry is because the setup of the socket (with internal_socket etc) might be incompatible with some of the changes. I see some lines of code that was added that made me think that you took this into consideration, but it is worth confirming by testing it (even manually is Ok). I am also OK with this being a separate fix PR.


We manually tested this using the web-mpc demo, led to a couple of changes but everything seems to work as intended now. Reminds me of a previous issue about having a server participate in multiple computations simultaneously, which JIFF by default and this extension don't support - but that is certainly  a separate PR

**Pull request merged**